rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/LocalInfrastructure.java

```diff
@@ -327,29 +330,42 @@ protected void initializePersistedInfraVariables() {
         this.persistedInfraVariables.put(LAST_NODE_STARTED_INDEX_KEY, 0);
     }
 
-    private void removeNodeAndShutdownProcessIfNeeded(String nodeUrl) {
+    private void removeNodeAndShutdownProcessIfNeeded(String nodeUrlToRemove) {
 
         // Update handle & acquire/lost nodes persistent counters
         decrementNumberOfHandledNodesWithLockAndPersist();
-        if (nodeSource.getNodeInDeployingOrLostNodes(nodeUrl) == null) {
-            decrementNumberOfLostNodesWithLockAndPersist();
-        } else {
+        if (nodeSource.getNodeInDeployingOrLostNodes(nodeUrlToRemove) == null) {
             decrementNumberOfAcquiredNodesWithLockAndPersist();
+        } else {
+            decrementNumberOfLostNodesWithLockAndPersist();
         }
 
-        // Delete the process if it doesn't have remaining node
         Iterator<Map.Entry<ProcessExecutor, List<String>>> processIterator = processExecutors.entrySet().iterator();
         while (processIterator.hasNext()) {
             Map.Entry<ProcessExecutor, List<String>> processExecutor = processIterator.next();
-            if (processExecutor.getValue().contains(nodeUrl)) {
-                processExecutor.getValue().remove(nodeUrl);
-                // if there is no remaining node, kill the JVM process
+            // Remove the nodeUrl if present
+            Iterator<String> nodesIterator = processExecutor.getValue().iterator();
+            boolean nodeFound = false;
+            while (nodesIterator.hasNext()) {
+                String nodeUrl = nodesIterator.next();
+                String nodeName = nodeUrl.substring(nodeUrl.lastIndexOf("/"));
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Put single-quotes around '/' to use the faster "lastIndexOf(char)" method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3027)